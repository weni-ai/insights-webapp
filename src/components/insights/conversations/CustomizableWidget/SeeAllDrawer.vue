<template>
  <UnnnicDrawer
    class="see-all-drawer"
    data-testid="see-all-drawer"
    :modelValue="modelValue"
    title="CUSTOM"
    size="lg"
    @close="emit('update:modelValue', false)"
  >
    <template #content>
      <ProgressTable
        :progressItems="formattedData"
        data-testid="see-all-drawer-table"
      />
    </template>
  </UnnnicDrawer>
</template>

<script setup lang="ts">
import ProgressTable from '@/components/ProgressTable.vue';
import { useWidgetFormatting } from '@/composables/useWidgetFormatting';
import type { CsatResult } from '@/services/api/resources/conversational/widgets';

const { formatPercentage, formatNumber } = useWidgetFormatting();

const props = defineProps<{
  modelValue: boolean;
  data: CsatResult[];
}>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const defaultColor = '#3182CE';
const defaultBackgroundColor = '#BEE3F8';

const formattedData = props.data.map((item) => ({
  label: item.label,
  value: item.value,
  description: `${formatPercentage(item.value)} (${formatNumber(item.full_value)})`,
  backgroundColor: defaultBackgroundColor,
  color: defaultColor,
  subItems: [],
}));
</script>
