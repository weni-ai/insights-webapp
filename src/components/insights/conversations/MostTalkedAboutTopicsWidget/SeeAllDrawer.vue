<template>
  <UnnnicDrawer
    class="see-all-drawer"
    :modelValue="modelValue"
    :title="$t('conversations_dashboard.most_talked_about_topics.title')"
    size="lg"
    @close="emit('update:modelValue', false)"
  >
    <template #content>
      <ProgressTable
        :progressItems="MOCK_DATA_PROGRESS_TABLE"
        :subItemsDescription="
          $t('conversations_dashboard.most_talked_about_topics.subtopics')
        "
        :expandedItems="expandedItems"
      />
    </template>
  </UnnnicDrawer>
</template>

<script setup lang="ts">
import ProgressTable from '@/components/ProgressTable.vue';
import { addColors } from '@/utils/treemap';
import { useWidgetFormatting } from '@/composables/useWidgetFormatting';

const { formatPercentage, formatNumber } = useWidgetFormatting();

defineProps<{
  modelValue: boolean;
}>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const MOCK_DATA = [
  {
    label: 'Entrega atrasada',
    value: 6973,
    percentage: 29,
    subItems: [
      {
        label: 'Atraso no transporte',
        value: 5229,
        percentage: 75,
      },
      {
        label: 'Problema na preparação',
        value: 1000,
        percentage: 25,
      },
    ],
  },
  {
    label: 'Produto defeituoso',
    value: 5500,
    percentage: 23,
  },
  {
    label: 'Dúvidas sobre preço',
    value: 1600,
    percentage: 16,
    subItems: [
      {
        label: 'Desconto no valor',
        value: 1825,
        percentage: 35,
      },
      {
        label: 'Desconto no frete',
        value: 375,
        percentage: 65,
      },
    ],
  },
  {
    label: 'Cancelamento',
    value: 1400,
    percentage: 14,
  },
  {
    label: 'Unclassified',
    value: 1000,
    percentage: 13,
  },
  {
    label: 'Outros',
    value: 500,
    percentage: 5,
  },
];

const coloredData = addColors(MOCK_DATA);

const MOCK_DATA_PROGRESS_TABLE = coloredData.map((item) => ({
  label: item.label,
  value: item.percentage,
  description: `${formatPercentage(item.percentage)} (${formatNumber(item.value)})`,
  backgroundColor: item.color,
  color: item.hoverColor,
  subItems: MOCK_DATA.find(
    (mockItem) => mockItem.label === item.label,
  )?.subItems?.map((subItem) => ({
    label: subItem.label,
    value: subItem.percentage,
    description: `${formatPercentage(subItem.percentage)} (${formatNumber(subItem.value)})`,
    backgroundColor: item.color,
    color: item.hoverColor,
  })),
}));

const expandedItems = defineModel<string[]>('expandedItems', {
  default: () => [],
});
</script>
