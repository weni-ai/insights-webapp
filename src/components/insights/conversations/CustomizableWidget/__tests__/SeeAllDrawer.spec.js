import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import SeeAllDrawer from '../SeeAllDrawer.vue';

const mockWidgetFormatting = {
  formatPercentage: vi.fn((value) => `${value}%`),
  formatNumber: vi.fn((value) => value.toString()),
};

vi.mock('@/composables/useWidgetFormatting', () => ({
  useWidgetFormatting: () => mockWidgetFormatting,
}));

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: { en: {} },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('SeeAllDrawer', () => {
  let wrapper;

  const defaultProps = {
    modelValue: true,
    data: [
      { label: 'Satisfied', value: 75.5, full_value: 302 },
      { label: 'Neutral', value: 15.2, full_value: 61 },
      { label: 'Unsatisfied', value: 9.3, full_value: 37 },
    ],
  };

  const createWrapper = (props = {}) =>
    mount(SeeAllDrawer, {
      props: { ...defaultProps, ...props },
      global: {
        stubs: {
          UnnnicDrawer: {
            template:
              '<div data-testid="see-all-drawer" class="see-all-drawer"><slot name="content" /></div>',
          },
          ProgressTable: true,
        },
      },
    });

  const drawer = () => wrapper.find('[data-testid="see-all-drawer"]');
  const progressTable = () =>
    wrapper.find('[data-testid="see-all-drawer-table"]');

  beforeEach(() => {
    vi.clearAllMocks();
    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render all elements', () => {
      expect(drawer().exists()).toBe(true);
      expect(progressTable().exists()).toBe(true);
    });

    it('should render with different data sets', () => {
      const customData = [
        { label: 'High', value: 90, full_value: 450 },
        { label: 'Low', value: 10, full_value: 50 },
      ];
      wrapper = createWrapper({ data: customData });

      expect(wrapper.props('data')).toEqual(customData);
      expect(progressTable().exists()).toBe(true);
    });

    it('should render with empty data', () => {
      wrapper = createWrapper({ data: [] });

      expect(wrapper.props('data')).toEqual([]);
      expect(progressTable().exists()).toBe(true);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(drawer().classes()).toContain('see-all-drawer');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Event handling', () => {
    it('should emit update:modelValue when closed', () => {
      wrapper.vm.$emit('update:modelValue', false);

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });

    it('should emit update:modelValue with correct value', () => {
      wrapper.vm.$emit('update:modelValue', true);

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([true]);
    });
  });

  describe('Data formatting', () => {
    it('should format data correctly', () => {
      const expectedFormattedData = [
        {
          label: 'Satisfied',
          value: 75.5,
          description: '75.5% (302)',
          backgroundColor: '#BEE3F8',
          color: '#3182CE',
          subItems: [],
        },
        {
          label: 'Neutral',
          value: 15.2,
          description: '15.2% (61)',
          backgroundColor: '#BEE3F8',
          color: '#3182CE',
          subItems: [],
        },
        {
          label: 'Unsatisfied',
          value: 9.3,
          description: '9.3% (37)',
          backgroundColor: '#BEE3F8',
          color: '#3182CE',
          subItems: [],
        },
      ];

      expect(wrapper.vm.formattedData).toEqual(expectedFormattedData);
    });

    it('should call formatting functions correctly', () => {
      wrapper.vm.formattedData;

      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(75.5);
      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(15.2);
      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(9.3);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(302);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(61);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(37);
    });

    it('should pass formatted data to ProgressTable', () => {
      expect(progressTable().attributes('progressitems')).toBeDefined();
    });
  });

  describe('Props validation', () => {
    const propTestCases = [
      { modelValue: true, data: defaultProps.data },
      { modelValue: false, data: [] },
      {
        modelValue: true,
        data: [{ label: 'Single', value: 100, full_value: 1000 }],
      },
    ];

    propTestCases.forEach(({ modelValue, data }) => {
      it(`should handle props: modelValue=${modelValue} with ${data.length} items`, () => {
        wrapper = createWrapper({ modelValue, data });

        expect(wrapper.props('modelValue')).toBe(modelValue);
        expect(wrapper.props('data')).toEqual(data);
      });
    });
  });
});
