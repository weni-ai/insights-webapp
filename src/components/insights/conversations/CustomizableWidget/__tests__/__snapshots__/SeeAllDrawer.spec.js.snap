// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SeeAllDrawer > Component structure > should match snapshot 1`] = `
<div
  class="see-all-drawer"
  data-testid="see-all-drawer"
  modelvalue="true"
  size="lg"
  title="CUSTOM"
>
  
  <progress-table-stub
    data-testid="see-all-drawer-table"
    isloading="false"
    progressitems="[object Object],[object Object],[object Object]"
  />
  
</div>
`;
